File: typescript/tests/_generation.ts
import type { MemoryEngine } from '../src/testing/memory.js'
import { EPOCH_COLUMN, FENCE_PREFIX, WriteFence } from '../src/write-fence.js'
import { MemoryFences } from './_write-fence.js'

export function bindGenerationMetadata(engines: Record<string, MemoryEngine>, metadata:
  Record<string, Record<string, { project_id: string; epoch: number }>>): void {
  for (const [name, tables] of Object.entries(metadata)) {
    const backends = new Map<string, MemoryFences>()
    for (const [table, record] of Object.entries(tables)) {
      const backend = new MemoryFences()
      backend.identity = name + '/' + table
      backend.column = 'valid'
      backend.constraints = {
        [FENCE_PREFIX + 'owner_' + record.project_id]: '1',
        [FENCE_PREFIX + 'min_' + record.epoch]: `${EPOCH_COLUMN} >= ${record.epoch}`,
        [FENCE_PREFIX + 'max_' + record.epoch]: `${EPOCH_COLUMN} <= ${record.epoch}`,
      }
      backends.set(table, backend)
    }
    Object.defineProperty(engines[name], 'validateSchema', { value: async () => undefined })
    Object.defineProperty(engines[name], 'writeFence', {
      value: (table: string, options: { projectId: string }) => new WriteFence(backends.get(table)!, table, options),
    })
  }
}

File: python/tests/_generation.py
"""Generation metadata for shared vectors; rows/calls use the existing SDK memory engine."""

from __future__ import annotations

from typing import Any

from _write_fence import MemoryFences

from sde.write_fence import EPOCH_COLUMN, FENCE_PREFIX, WriteFence


def bind_generation_metadata(engines: dict[str, Any], metadata: dict[str, Any]) -> None:
    for name, tables in metadata.items():
        backends: dict[str, MemoryFences] = {}
        for table, record in tables.items():
            backend = MemoryFences()
            backend.identity = name + "/" + table
            backend.column = "valid"
            epoch = record["epoch"]
            backend.constraints = {
                FENCE_PREFIX + "owner_" + record["project_id"]: "1",
                FENCE_PREFIX + "min_" + str(epoch): EPOCH_COLUMN + " >= " + str(epoch),
                FENCE_PREFIX + "max_" + str(epoch): EPOCH_COLUMN + " <= " + str(epoch),
            }
            backends[table] = backend

        def factory(
            table: str, *, project_id: str, backends: dict[str, MemoryFences] = backends
        ) -> WriteFence:
            return WriteFence(backends[table], table, project_id=project_id)

        engines[name].write_fence = factory
        engines[name].validate_schema = lambda _layout: None
